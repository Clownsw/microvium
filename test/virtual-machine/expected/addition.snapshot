

slot ['dummy.mvms:#entry'] = &function ['dummy.mvms:#entry'];
slot ['dummy.mvms:exports'] = &allocation 1;

function ['dummy.mvms:#entry']() {
  entry:
    LoadArg(index 0);
    StoreGlobal(name 'dummy.mvms:exports');
    Literal(lit 1);
    Literal(lit 2);
    BinOp(op '+');
    Pop(count 1);
    Literal(lit undefined);
    Return();
}

allocation 1 = {
};