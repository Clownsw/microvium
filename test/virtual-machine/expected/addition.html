
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      </head>
      <body>
      
  <table class="visual-buffer">
    <colgroup>
      <col>
      <col>
      <col>
      <col>
    </colgroup>
    <tbody>
      
      <tr>
        <td class="address">
          <a class="address-text" id="address0000" href="#address0000">
            0000
          </a>
        </td>
        <td class="data">
          <span class="byte">04</span>
        </td>
        <td class="label">
          bytecodeVersion: 
        </td>
        <td class="value">4</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0001" href="#address0001">
            0001
          </a>
        </td>
        <td class="data">
          <span class="byte">1C</span>
        </td>
        <td class="label">
          headerSize: 
        </td>
        <td class="value">28</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0002" href="#address0002">
            0002
          </a>
        </td>
        <td class="data">
          <span class="byte">03</span>
        </td>
        <td class="label">
          requiredEngineVersion: 
        </td>
        <td class="value">3</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0003" href="#address0003">
            0003
          </a>
        </td>
        <td class="data">
          <span class="byte">00</span>
        </td>
        <td class="label">
          reserved: 
        </td>
        <td class="value">0</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0004" href="#address0004">
            0004
          </a>
        </td>
        <td class="data">
          <span class="byte">48</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          bytecodeSize: 
        </td>
        <td class="value">72</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0006" href="#address0006">
            0006
          </a>
        </td>
        <td class="data">
          <span class="byte">E0</span><wbr><span class="byte">E6</span>
        </td>
        <td class="label">
          crc: 
        </td>
        <td class="value">0xE6E0</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0008" href="#address0008">
            0008
          </a>
        </td>
        <td class="data">
          <span class="byte">03</span><wbr><span class="byte">00</span><wbr><span class="byte">00</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          requiredFeatureFlags: 
        </td>
        <td class="value">0x00000003</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address000C" href="#address000C">
            000C
          </a>
        </td>
        <td class="data">
          <span class="byte">1C</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_IMPORT_TABLE]: 
        </td>
        <td class="value">0x001C</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address000E" href="#address000E">
            000E
          </a>
        </td>
        <td class="data">
          <span class="byte">1C</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_EXPORT_TABLE]: 
        </td>
        <td class="value">0x001C</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0010" href="#address0010">
            0010
          </a>
        </td>
        <td class="data">
          <span class="byte">1C</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_SHORT_CALL_TABLE]: 
        </td>
        <td class="value">0x001C</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0012" href="#address0012">
            0012
          </a>
        </td>
        <td class="data">
          <span class="byte">1C</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_BUILTINS]: 
        </td>
        <td class="value">0x001C</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0014" href="#address0014">
            0014
          </a>
        </td>
        <td class="data">
          <span class="byte">20</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_STRING_TABLE]: 
        </td>
        <td class="value">0x0020</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0016" href="#address0016">
            0016
          </a>
        </td>
        <td class="data">
          <span class="byte">22</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_ROM]: 
        </td>
        <td class="value">0x0022</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0018" href="#address0018">
            0018
          </a>
        </td>
        <td class="data">
          <span class="byte">3A</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_GLOBALS]: 
        </td>
        <td class="value">0x003A</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address001A" href="#address001A">
            001A
          </a>
        </td>
        <td class="data">
          <span class="byte">3E</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          sectionOffsets[BCS_HEAP]: 
        </td>
        <td class="value">0x003E</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address001C" href="#address001C">
            001C
          </a>
        </td>
        <td class="data">
          <span class="byte">09</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          builtin[BIN_INTERNED_STRINGS]: 
        </td>
        <td class="value">0x0009</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address001E" href="#address001E">
            001E
          </a>
        </td>
        <td class="data">
          <span class="byte">3D</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          builtin[BIN_ARRAY_PROTO]: 
        </td>
        <td class="value">0x003D</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0020" href="#address0020">
            0020
          </a>
        </td>
        <td class="data">
          <span class="byte">25</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          &push: 
        </td>
        <td class="value">0x0025</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0022" href="#address0022">
            0022
          </a>
        </td>
        <td class="data">
          <span class="byte">05</span><wbr><span class="byte">40</span>
        </td>
        <td class="label">
          Allocation header: 
        </td>
        <td class="value">0x4005</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0024" href="#address0024">
            0024
          </a>
        </td>
        <td class="data">
          <span class="byte">70</span><wbr><span class="byte">75</span><wbr><span class="byte">73</span><wbr><span class="byte">68</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          String: 
        </td>
        <td class="value">&#39;push&#39;</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0029" href="#address0029">
            0029
          </a>
        </td>
        <td class="data">
          <span class="byte">00</span>
        </td>
        <td class="label">
          
        </td>
        <td class="value">&lt;padding&gt;</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address002A" href="#address002A">
            002A
          </a>
        </td>
        <td class="data">
          <span class="byte">0D</span><wbr><span class="byte">50</span>
        </td>
        <td class="label">
          Func alloc header (Array_push): 
        </td>
        <td class="value">0x500D</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address002C" href="#address002C">
            002C
          </a>
        </td>
        <td class="data">
          <span class="byte">04</span>
        </td>
        <td class="label">
          maxStackDepth: 
        </td>
        <td class="value">4</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address002D" href="#address002D">
            002D
          </a>
        </td>
        <td class="data">
          <span class="byte">31</span>
        </td>
        <td class="label">
          VM_OP_LOAD_ARG_1(0x1): 
        </td>
        <td class="value">LoadArg(index 1)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address002E" href="#address002E">
            002E
          </a>
        </td>
        <td class="data">
          <span class="byte">30</span>
        </td>
        <td class="label">
          VM_OP_LOAD_ARG_1(0x0): 
        </td>
        <td class="value">LoadArg(index 0)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address002F" href="#address002F">
            002F
          </a>
        </td>
        <td class="data">
          <span class="byte">30</span>
        </td>
        <td class="label">
          VM_OP_LOAD_ARG_1(0x0): 
        </td>
        <td class="value">LoadArg(index 0)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0030" href="#address0030">
            0030
          </a>
        </td>
        <td class="data">
          <span class="byte">88</span><wbr><span class="byte">21</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          VM_OP_EXTENDED_3(VM_OP3_LOAD_LITERAL): 
        </td>
        <td class="value">Literal(lit &#39;length&#39;)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0033" href="#address0033">
            0033
          </a>
        </td>
        <td class="data">
          <span class="byte">6B</span>
        </td>
        <td class="label">
          VM_OP_EXTENDED_1(VM_OP1_OBJECT_GET_1): 
        </td>
        <td class="value">ObjectGet()</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0034" href="#address0034">
            0034
          </a>
        </td>
        <td class="data">
          <span class="byte">12</span>
        </td>
        <td class="label">
          VM_OP_LOAD_VAR_1(0x2): 
        </td>
        <td class="value">LoadVar(index 0)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0035" href="#address0035">
            0035
          </a>
        </td>
        <td class="data">
          <span class="byte">6F</span>
        </td>
        <td class="label">
          VM_OP_EXTENDED_1(VM_OP1_OBJECT_SET_1): 
        </td>
        <td class="value">ObjectSet()</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0036" href="#address0036">
            0036
          </a>
        </td>
        <td class="data">
          <span class="byte">67</span>
        </td>
        <td class="label">
          VM_OP_EXTENDED_1(VM_OP1_POP): 
        </td>
        <td class="value">Pop(count 1)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0037" href="#address0037">
            0037
          </a>
        </td>
        <td class="data">
          <span class="byte">01</span>
        </td>
        <td class="label">
          VM_OP_LOAD_SMALL_LITERAL(0x1): 
        </td>
        <td class="value">Literal(lit undefined)</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0038" href="#address0038">
            0038
          </a>
        </td>
        <td class="data">
          <span class="byte">60</span>
        </td>
        <td class="label">
          VM_OP_EXTENDED_1(VM_OP1_RETURN): 
        </td>
        <td class="value">Return()</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0039" href="#address0039">
            0039
          </a>
        </td>
        <td class="data">
          <span class="byte">00</span>
        </td>
        <td class="label">
          
        </td>
        <td class="value">&lt;padding&gt;</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address003A" href="#address003A">
            003A
          </a>
        </td>
        <td class="data">
          <span class="byte">1D</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          
        </td>
        <td class="value">&lt;pad with deleted&gt;</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address003C" href="#address003C">
            003C
          </a>
        </td>
        <td class="data">
          <span class="byte">02</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          Handle: 
        </td>
        <td class="value">0x0002</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address003E" href="#address003E">
            003E
          </a>
        </td>
        <td class="data">
          <span class="byte">08</span><wbr><span class="byte">C0</span>
        </td>
        <td class="label">
          TsPropertyList.[header]: 
        </td>
        <td class="value">0xC008</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0040" href="#address0040">
            0040
          </a>
        </td>
        <td class="data">
          <span class="byte">09</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          TsPropertyList.dpNext: 
        </td>
        <td class="value">0x0009</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0042" href="#address0042">
            0042
          </a>
        </td>
        <td class="data">
          <span class="byte">09</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          TsPropertyList.dpProto: 
        </td>
        <td class="value">0x0009</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0044" href="#address0044">
            0044
          </a>
        </td>
        <td class="data">
          <span class="byte">25</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          TsPropertyList.keys[0]: 
        </td>
        <td class="value">0x0025</td>
      </tr>

      <tr>
        <td class="address">
          <a class="address-text" id="address0046" href="#address0046">
            0046
          </a>
        </td>
        <td class="data">
          <span class="byte">2D</span><wbr><span class="byte">00</span>
        </td>
        <td class="label">
          TsPropertyList.values[push]: 
        </td>
        <td class="value">0x002D</td>
      </tr>
      
      <tr>
        <td class="address">
          <a class="address-text" id="address0048" href="#address0048">
            0048
          </a>
        </td>
        <td class="data">
          
        </td>
        <td class="label">
          
        </td>
        <td class="value"></td>
      </tr>
    </tbody>
  </table>
      <style>body {
  font-family: sans-serif;
  margin: 20px;
}

table.visual-buffer {
  border-collapse: collapse;
  table-layout: fixed;
  width: 1000px;
  border-spacing: 0;
  font-size: 16px;
  font-family: 'Consolas', 'Courier New', Courier, monospace;
}

table.visual-buffer td {
  vertical-align: top;
  padding: 2px 0;
  margin: 0;
}

/* ---------------- Column 1: Address -------------------*/
table.visual-buffer col:nth-child(1) {
  width: 47px;
}

table.visual-buffer td:nth-child(1) {
  text-align: left;
  padding-right: 8px;
  border-top: 1px solid rgb(100,100,100);
}

.visual-buffer .address-text {
  font-size: 85%;
  /* font-weight: bold; */
  font-family: 'Consolas', 'Courier New', Courier, monospace;
  position: relative;
  top: -12px;
  border-radius: 2px;
  background-color: white;
  padding: 1px 4px;
  color: black;
  text-decoration: none;
}

.visual-buffer .address-text:hover {
  color: rgb(40, 40, 80);
  background-color: rgb(220, 220, 240);
}

/* ---------------- Column 2: Binary -------------------*/
table.visual-buffer col:nth-child(2) {
  background: rgb(40,40,50);
  width: 141px;
}

table.visual-buffer td:nth-child(2) {
  text-align: left;
  padding-left: 4px;
  padding-right: 4px;
  border-top: 1px solid rgba(255, 255, 255, 0.06);
  font-size: 0;
  height: 0px;
  overflow: hidden;
}

.visual-buffer .data .byte {
  display: inline-block;
  box-sizing: border-box;
  width: 29px;
  height: 29px;
  font-family: 'Consolas', 'Courier New', Courier, monospace;
  font-size: 17px;
  color: white;
  background: rgb(60,60,75);
  border-radius: 2px;
  text-align: center;
  padding-top: 6px;
  margin: 2px 2px;
  overflow: hidden;
}

.visual-buffer .data .byte.pending {
  box-shadow: 0px 0px 5px rgba(255, 255, 255, 0.3) inset;
  border-radius: 3px;
}

/* ---------------- Column 3: Label -------------------*/
table.visual-buffer col:nth-child(3) {
  width: 350px;
}

table.visual-buffer td:nth-child(3) {
  text-align: right;
  padding-right: 5px;
  opacity: 0.7;
  border-top: 1px solid rgba(100, 100, 100, 0.25);
  padding-top: 10px;
}

/* ---------------- Column 4: Value -------------------*/
table.visual-buffer col:nth-child(4) {
  text-align: left;
}

table.visual-buffer td:nth-child(4) {
  border-top: 1px solid rgba(100, 100, 100, 0.25);
  padding-top: 10px;
}

.visual-buffer .pending-value {
  font-style: italic;
  opacity: 0.5;
}
</style>
      </body>
      </html>